### PR TITLE
feat: make the aws tf code more modular

### PR DESCRIPTION
**Pull request by frezbo**
Make the aws code more modular so that it can be used in CI.

### CODE REVIEW

**Comment by frezbo** [comment]
An example usage with a var file:

```json
{
    "cluster_name": "talos-nvidia-test",
    "num_control_planes": 1,
    "num_workers": 0,
    "ami_id": "ami-034f35c36088696a8",
    "instance_type_control_plane": "t3.medium",
    "config_patch_files_worker": [
        "patch.yaml"
    ],
    "extra_tags": {
        "Project": "talos-nvidia-test",
        "Environment": "ci test",
        "Owner": "frezbo"
    },
    "node_groups": [
        {
            "name": "nvidia-t4",
            "num_instances": 2,
            "instance_type": "g4dn.xlarge",
            "tags": {
                "Type": "nvidia-t4"
            }
        },
        {
            "name": "nvidia-a100",
            "num_instances": 1,
            "instance_type": "p4d.24xlarge",
            "tags": {
                "Type": "nvidia-a100"
            },
            "config_patch_files": [
                "patch-a100.yaml"
            ]
        }
    ]
}
```

**patch.yaml**

```yaml
machine:
  kernel:
    modules:
      - name: nvidia
      - name: nvidia_uvm
      - name: nvidia_drm
      - name: nvidia_modeset
  sysctls:
    net.core.bpf_jit_harden: 1
  install:
    extensions:
      - image: ghcr.io/frezbo/nvidia-container-toolkit:535.54.03-v1.13.5
      - image: ghcr.io/frezbo/nvidia-open-gpu-kernel-modules:535.54.03-v1.5.0-alpha.3-2-gc59245d-dirty
```

**patch-a100.yaml**

```yaml
machine:
  install:
    extensions:
      - image: ghcr.io/frezbo/nvidia-fabricmanager:535.54.03
```

File: examples/terraform/aws/main.tf [hunk 1]
```diff
@@ -132,10 +148,33 @@ module "talos_worker_nodes" {
...
+      volume_size = 100
+    }
+  ]
+}
+
+module "talos_worker_node_group" {
+  source  = "terraform-aws-modules/ec2-instance/aws"
+  version = "~> 4.0"
+
+  for_each = merge([for info in var.node_groups : { for index in range(0, info.num_instances) : "${info.name}.${index}" => info }]...)
```
Comment (frezbo) [inline examples/terraform/aws/main.tf:170]: don't ask me how I came up with this :sweat_smile:

**Review by smira — APPROVED** [review]
I'm not a TF expert, but I don't see anything concerning :)

File: examples/terraform/aws/main.tf [hunk 2]
```diff
@@ -155,9 +197,11 @@ data "talos_machine_configuration" "controlplane" {
   machine_secrets  = talos_machine_secrets.this.machine_secrets
   docs             = false
   examples         = false
-  config_patches = [
-    yamlencode(local.common_machine_config_patch)
-  ]
+  config_patches = concat(
+    local.config_patches_common,
```
Comment (rsmitty) [inline examples/terraform/aws/main.tf:201]: Shouldn't all patches be run through the yamlencode or am I missing something?

Comment (rsmitty) [reply examples/terraform/aws/main.tf:201]: Oh, nvm, I see the yamlencode is specific just for registering kubelet with fqdn

File: examples/terraform/aws/variables.tf [hunk 3]
```diff
@@ -16,12 +16,59 @@ variable "num_workers" {
...
 }
 
+variable "instance_type_worker" {
+  description = "Instance type to use for the worker nodes"
+  type        = string
+  default     = "c5.large"
+}
+
+variable "iam_instance_profile_control_plane" {
+  description = "IAM instance profile to use for the control plane nodes, if not set an IAM profile will not be attached"
```
Comment (rsmitty) [inline examples/terraform/aws/variables.tf:42]: Are there known requirements for the IAM group? If so, we should at least provide a link as to what should be configured if it's going to have to be pre-created.

**Comment by rsmitty** [comment]
Generally fine with this, but I think that calling them "node groups" is a bit confusing, being that that's a construct of EKS. "worker groups" feels more appropriate imo.

File: examples/terraform/aws/variables.tf (hunk 3 above)
Comment (frezbo) [reply examples/terraform/aws/variables.tf:42]: probably only if deploying a CCM, I'll find some links, for now I want to integrate this to CI :sweat_smile:

**Comment by frezbo** [comment]
> 

makes sense I'll update

**Review by rsmitty — APPROVED** [review]

**Pull request merged**